### PR TITLE
New direction

### PR DESCRIPTION
**Pull request by lwahlmeier**
Image is now just an interface that different backends can use.  There is currently a JavaImage (pure java implementation) and an AWTImage (using java.awt standard CP).  The plan is to write more extensions using OpenGL and SWT.

### CODE REVIEW

File: src/main/java/org/java_lcw/jil/AWTImage.java [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+      nbi = new BufferedImage(width, height, BufferedImage.TYPE_4BYTE_ABGR);
+    }
+    nbi.getGraphics().drawImage(bi, 0, 0, null);
+    return new AWTImage(nbi, MODE);
+  }
+
+  @Override
+  public AWTImage resizeWithBorders(int bWidth, int bHeight, Color borderColor,
+      ScaleType st) {
+    return null;
```
Comment (jentfoo) [inline src/main/java/org/java_lcw/jil/AWTImage.java]: Seems like this should throw an UnsupportedOperationException?

File: src/main/java/org/java_lcw/jil/AWTImage.java [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+
+    return nimg;
+  }
+
+  @Override
+  public void fillImageWithColor(Color c) {
+    Graphics2D graphics = bi.createGraphics();
+    graphics.setPaint(Utils.toAWTColor(c));
+    graphics.fillRect(0, 0, width, height);
+    graphics.dispose();
```
Comment (jentfoo) [inline src/main/java/org/java_lcw/jil/AWTImage.java]: Should this dispose be in a try/finally block?

File: src/main/java/org/java_lcw/jil/AWTImage.java [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+    if(img instanceof AWTImage) {
+      bi2 = ((AWTImage)img).bi;
+    } else {
+      bi2 = AWTImage.fromJavaImage(img.toJavaImage()).bi;
+    }
+    Graphics2D gi = bi.createGraphics();
+    float opacity = 1.0f;
+    gi.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC, opacity));
+    gi.drawImage(bi2, x, y, null);
+    gi.dispose();
```
Comment (jentfoo) [inline src/main/java/org/java_lcw/jil/AWTImage.java]: I see there are other cases of this...same as above. (but I will stop commenting on them, so may need to search for dispose instances)

File: src/main/java/org/java_lcw/jil/Draw.java [hunk 4]
```diff
@@ -1,250 +1,19 @@
...
-        } 
-      }
-    }
-    if(fill) {
-      int tx = x+ w/2;
-      int ty = y+ h/2;
-      floodFill(img, tx, ty, c, c, false);
-    }
-  }
+  public void drawRect(int x, int y, int w, int h, Color c, int lineWidth, boolean fill);
```
Comment (jentfoo) [inline src/main/java/org/java_lcw/jil/Draw.java]: Worth adding javadocs?

File: src/main/java/org/java_lcw/jil/JavaImage.java [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+import org.java_lcw.jil.scalers.NearestNeighborScaler;
+
+
+/**
+ *
+ * Main Image object used to construct new Image files.  
+ * 
+ * @author lcw - Luke Wahlmeier
+ */
+public class JavaImage implements Image {
```
Comment (jentfoo) [inline src/main/java/org/java_lcw/jil/JavaImage.java]: A possible better name than JavaImage would be "JilImage" (though typing it, the l and capital I looks weird next to each other).

Technically AWT is a "java" Image as well.  I feel like "JilImage" better describes what it is.

**Comment by jentfoo** [comment]
General comment...

Some things are better in the "JavaImage", while woulden't some things be better (more efficent) in the "AwtImage".  Do you basically expect the user to know when the right time to convert back and forth is?

Assuming the above is true, maybe it makes sense to have a "GeneralImage", or "PerformantImage", or something with a better name.  This implementation would be backed by a generic "Image", and before each operaiton it would do the conversion which makes sense (and obviously if it's already the right type that conversion is a no-op).  Then the new one would store that implementation inside it.

This way the user can let you choose what operations make the most sense with what image types, but you still have these raw image types for those that want to really be precise about what they are doing.

Just a thought, what do you think?

File: src/main/java/org/java_lcw/jil/AWTImage.java (hunk 1 above)
Comment (lwahlmeier) [inline src/main/java/org/java_lcw/jil/AWTImage.java]: Technically its not unsupported, I just didnt implement it... I will add an exception for now

File: src/main/java/org/java_lcw/jil/AWTImage.java (hunk 2 above)
Comment (lwahlmeier) [inline src/main/java/org/java_lcw/jil/AWTImage.java]: I dont think so, there is not really much here that can throw.  Also Graphics .dispose are only reallu an issue until the BufferedImage is GCed, then they are all disposed.

File: src/main/java/org/java_lcw/jil/Draw.java (hunk 4 above)
Comment (lwahlmeier) [inline src/main/java/org/java_lcw/jil/Draw.java]: for sure

File: src/main/java/org/java_lcw/jil/JavaImage.java (hunk 5 above)
Comment (lwahlmeier) [inline src/main/java/org/java_lcw/jil/JavaImage.java]: Will Do

**Comment by lwahlmeier** [comment]
Technically speaking nothing is "better" in one or the other.  In general AWTImage will be faster, but vary a bit based on the JVM implementation (not all circles are drawn the same or scaling), the Java/JilImage will be consistent regardless of VM but in general most operations will be slower.  

I except that people will go back and forth very rarely.  They will choose an implementation and just use that.  The Idea is that all Implementations *must* have a way to translate into a JILImage and from a JILImage allowing the user to move between formats if for some reason they want too.

The main idea is that JIL is a standard convenience wrapper to standardize the way java images are dealt with.  The implementation behind it does not really matter except to the person using it.

**Comment by lwahlmeier** [comment]
Basically did everything you suggested.  Not sure the try/finally for disposes was needed but I dont think it hurt (other then messy).

**Comment by jentfoo** [comment]
Looks good to me.  Want me to go ahead and merge it, or is there more you were planning to do?

**Comment by lwahlmeier** [comment]
Go ahead and merge it. :)

**Pull request merged**